### PR TITLE
Added backup options when creating/updating database

### PR DESCRIPTION
**Pull request by bastien70**
This PR should allow users to set backup options when creating or updating database

### CODE REVIEW

File: migrations/Version20220729140330.php [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public function getDescription(): string
+    {
+        return 'Add options for backups';
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE `database` ADD reset_auto_increment TINYINT(1) DEFAULT FALSE NOT NULL, ADD add_drop_database TINYINT(1) DEFAULT FALSE NOT NULL, ADD add_drop_table TINYINT(1) DEFAULT TRUE NOT NULL, ADD add_drop_trigger TINYINT(1) DEFAULT TRUE NOT NULL, ADD add_locks TINYINT(1) DEFAULT TRUE NOT NULL, ADD complete_insert TINYINT(1) DEFAULT FALSE NOT NULL');
```
Comment (jmsche) [inline migrations/Version20220729140330.php:23]:
```suggestion
        $this->addSql('ALTER TABLE `database` ADD reset_auto_increment TINYINT(1) DEFAULT FALSE NOT NULL, ADD add_drop_database TINYINT(1) DEFAULT FALSE NOT NULL, ADD add_drop_table TINYINT(1) DEFAULT TRUE NOT NULL, ADD add_drop_trigger TINYINT(1) DEFAULT TRUE NOT NULL, ADD add_locks TINYINT(1) DEFAULT TRUE NOT NULL, ADD complete_insert TINYINT(1) DEFAULT FALSE NOT NULL');
```

File: migrations/Version20220729140330.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    public function down(Schema $schema): void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE `database` DROP reset_auto_increment, DROP add_drop_database, DROP add_drop_table, DROP add_drop_trigger, DROP add_locks, DROP complete_insert');
```
Comment (jmsche) [inline migrations/Version20220729140330.php:29]:
```suggestion
        $this->addSql('ALTER TABLE `database` DROP reset_auto_increment, DROP add_drop_database, DROP add_drop_table, DROP add_drop_trigger, DROP add_locks, DROP complete_insert');
```

File: src/Controller/Admin/DatabaseCrudController.php [hunk 3]
```diff
@@ -258,5 +261,33 @@ public function configureFields(string $pageName): iterable
                 Database::STATUS_UNKNOWN => 'secondary',
             ])
             ->hideOnForm();
+
+        if (Crud::PAGE_INDEX !== $pageName) {
+            yield FormField::addPanel('database.panel.backup_options', 'fas fa-gear');
+
+            yield FormField::addRow();
```
Comment (jmsche) [inline src/Controller/Admin/DatabaseCrudController.php:268]: Is this one needed?

File: src/Controller/Admin/DatabaseCrudController.php [hunk 4]
```diff
@@ -258,5 +261,33 @@ public function configureFields(string $pageName): iterable
...
+                ->renderAsSwitch(false)
+                ->setColumns(6);
+
+            yield FormField::addRow();
+            yield BooleanField::new('addLocks', 'Add locks')
+                ->renderAsSwitch(false)
+                ->setColumns(6);
+            yield BooleanField::new('completeInsert', 'Complete insert')
+                ->renderAsSwitch(false)
+                ->setColumns(6);
```
Comment (jmsche) [inline src/Controller/Admin/DatabaseCrudController.php:290]: Even if we're not really translating options, these could still be translations

File: src/Entity/Database.php [hunk 5]
```diff
@@ -71,6 +71,24 @@ class Database implements \Stringable
...
+    private bool $addDropTable = true;
+
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $addDropTrigger = true;
+
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $addLocks = true;
+
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $completeInsert = false;
```
Comment (jmsche) [inline src/Entity/Database.php:90]: What about putting them in an embedded $options instead?

File: src/Entity/Database.php [hunk 6]
```diff
@@ -71,6 +71,24 @@ class Database implements \Stringable
...
     private ?AdapterConfig $adapter = null;
 
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $resetAutoIncrement = false;
+
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $addDropDatabase = false;
+
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $addDropTable = true;
```
Comment (jmsche) [inline src/Entity/Database.php:81]: This one is false by default

File: src/Service/BackupService.php [hunk 7]
```diff
@@ -163,9 +163,7 @@ private function defineMysqlDumpObject(Database $database): Mysqldump
             $database->getDsn(),
             $database->getUser(),
             $this->encryptor->decrypt($database->getPassword()),
-            [
-                'add-drop-table' => true,
-            ]
+            $database->getBackupOptions()
```
Comment (jmsche) [inline src/Service/BackupService.php:166]:
```suggestion
            $database->getBackupOptions(),
```

**Review by jmsche — CHANGES_REQUESTED** [review]

File: src/Controller/Admin/DatabaseCrudController.php (hunk 3 above)
Comment (bastien70) [reply src/Controller/Admin/DatabaseCrudController.php:268]: Yes options do not need to be on the index. Afterwards, I can override the condition by putting a ->hideOnIndex() on each field if you prefer

File: src/Entity/Database.php (hunk 6 above)
Comment (bastien70) [reply src/Entity/Database.php:81]: You had set this option to true by default before managing the options, hence the fact that I set it to true by default. (https://github.com/bastien70/dbsaver/blob/b7a5716a1c6f88288f9b91be0fc2ea90f9abb1e3/src/Service/BackupService.php#L160-L170) Do you still want me to set it to false by default?

File: src/Controller/Admin/DatabaseCrudController.php (hunk 3 above)
Comment (jmsche) [reply src/Controller/Admin/DatabaseCrudController.php:268]: I meant the addRow() in particular

File: src/Entity/Database.php (hunk 6 above)
Comment (jmsche) [reply src/Entity/Database.php:81]: Let's use the default from the package

File: src/Entity/Database.php (hunk 5 above)
Comment (bastien70) [reply src/Entity/Database.php:90]: Why not just a trait ?

File: src/Controller/Admin/DatabaseCrudController.php (hunk 3 above)
Comment (bastien70) [reply src/Controller/Admin/DatabaseCrudController.php:268]: Oh okay, no I can remove it

File: src/Entity/Database.php (hunk 5 above)
Comment (jmsche) [reply src/Entity/Database.php]: Because it's not at all the role of a trait

File: migrations/Version20220801084444.php [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace DoctrineMigrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20220801084444 extends AbstractMigration
```
Comment (jmsche) [inline migrations/Version20220801084444.php:13]:
```suggestion
final class Version20220801084444 extends AbstractMigration
```

File: src/Entity/Embed/Options.php [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $addDropTrigger = true;
+
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $addLocks = true;
+
+    #[ORM\Column(type: Types::BOOLEAN)]
+    private bool $completeInsert = false;
+
+    public function isResetAutoIncrement(): ?bool
```
Comment (jmsche) [inline src/Entity/Embed/Options.php:31]: All getters should return bool, not ?bool

**Review by jmsche — CHANGES_REQUESTED** [review]

**Review by jmsche — APPROVED** [review]

**Pull request merged**